File: app/lib/services/image_compress.dart
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:flutter_image_compress/flutter_image_compress.dart';
import 'package:path_provider/path_provider.dart';

/// Nén ảnh client-side trước khi upload.
///
/// Backend chỉ accept JPEG/PNG/WebP (cloudinary.service.ts ALLOWED_MIME).
/// Một số máy mới (Realme/Samsung/iPhone) chụp ảnh HEIC mặc định → phải convert.
///
/// Logic:
/// - Detect format qua magic bytes (header)
/// - Nếu đã là JPEG/PNG/WebP và < 500KB → skip (giữ chất lượng gốc)
/// - Còn lại (HEIC/HEIF/lớn) → re-encode sang JPEG quality 80, max 1920x1920
class ImageCompress {
  static const int _maxWidth = 1920;
  static const int _maxHeight = 1920;
  static const int _quality = 80;
  static const int _skipSizeThreshold = 500 * 1024; // 500KB

  /// Detect format qua magic bytes của file.
  /// Trả về true nếu là JPEG/PNG/WebP (backend accept trực tiếp).
  static bool _isAcceptedFormat(Uint8List bytes) {
    if (bytes.length < 12) return false;
    // JPEG: FF D8 FF
    if (bytes[0] == 0xFF && bytes[1] == 0xD8 && bytes[2] == 0xFF) return true;
    // PNG: 89 50 4E 47 0D 0A 1A 0A
    if (bytes[0] == 0x89 && bytes[1] == 0x50 && bytes[2] == 0x4E && bytes[3] == 0x47) return true;
    // WebP: RIFF....WEBP
    if (bytes[0] == 0x52 && bytes[1] == 0x49 && bytes[2] == 0x46 && bytes[3] == 0x46 &&
        bytes[8] == 0x57 && bytes[9] == 0x45 && bytes[10] == 0x42 && bytes[11] == 0x50) return true;
    return false;
  }

  /// Nén 1 ảnh từ XFile (image_picker).
  /// Trả về XFile mới (file tạm), giữ nguyên API với code cũ.
  static Future<XFile> compress(XFile original) async {
    final bytes = await original.readAsBytes();
    final isAccepted = _isAcceptedFormat(bytes);

    // Skip nén nếu đã đúng format VÀ nhỏ
    if (isAccepted && bytes.length < _skipSizeThreshold) return original;

    try {
      final compressed = await FlutterImageCompress.compressWithList(
        bytes,
        minWidth: _maxWidth,
        minHeight: _maxHeight,
        quality: _quality,
        format: CompressFormat.jpeg,
      );

      // Lưu file tạm với extension .jpg để mime detector backend nhận đúng image/jpeg
      final tmpDir = await getTemporaryDirectory();
      final tmpPath = '${tmpDir.path}/compressed_${DateTime.now().millisecondsSinceEpoch}.jpg';
      await File(tmpPath).writeAsBytes(compressed);

      if (kDebugMode) {
        final reduction = ((1 - compressed.length / bytes.length) * 100).toStringAsFixed(0);
        debugPrint('[ImageCompress] ${(bytes.length / 1024).toStringAsFixed(0)}KB → ${(compressed.length / 1024).toStringAsFixed(0)}KB (-$reduction%)');
      }

      return XFile(tmpPath);
    } catch (e) {
      debugPrint('[ImageCompress] error: $e');
      // Nếu input đã là JPEG/PNG/WebP thì fallback ảnh gốc OK (backend accept)
      // Nếu input là HEIC mà nén fail → throw để UI báo lỗi rõ ràng thay vì để
      // backend reject với message khó hiểu cho user
      if (isAccepted) return original;
      throw Exception('Không xử lý được ảnh (định dạng HEIC/HEIF không hỗ trợ trên thiết bị này). Vui lòng đổi cài đặt camera sang JPEG.');
    }
  }

  /// Nén loạt ảnh TUẦN TỰ (không phải song song).
  ///
  /// TM6 (Tier 2 audit): trước đây dùng `Future.wait(...)` chạy parallel:
  /// 10 ảnh × 5MB load đồng thời vào memory → 50MB+ uncompressed in-flight.
  /// Trên máy yếu (Galaxy J6, Redmi 8, 1-2GB RAM) → OOM crash app trước khi
  /// nén xong. Sequential thì chậm hơn ~2-3x nhưng KHÔNG crash.
  ///
  /// Trade-off chấp nhận:
  /// - Mạng tốt + máy mạnh: chậm hơn 2-3 giây cho 10 ảnh — không đáng kể
  /// - Mạng yếu + máy yếu: KHÔNG crash, user upload thành công thay vì mất bài
  ///
  /// Optional callback `onProgress(current, total)` để UI hiện progress bar.
  static Future<List<XFile>> compressBatch(
    List<XFile> originals, {
    void Function(int current, int total)? onProgress,
  }) async {
    final result = <XFile>[];
    for (var i = 0; i < originals.length; i++) {
      onProgress?.call(i, originals.length);
      result.add(await compress(originals[i]));
    }
    onProgress?.call(originals.length, originals.length);
    return result;
  }
}

File: web/lib/imageCompress.ts
/**
 * Nén ảnh client-side trước khi upload — mirror app/lib/services/image_compress.dart
 * để web không gửi ảnh gốc (5-12MB từ điện thoại đời mới) gây chậm 5s+ khi đăng bài.
 *
 * Resize tối đa 1920x1920, JPEG quality 0.8. Skip nếu ảnh đã nhỏ (<500KB) và đúng
 * format JPEG/PNG/WebP — giữ chất lượng gốc, không nén lại vô ích.
 */

const MAX_DIMENSION = 1920;
const JPEG_QUALITY = 0.8;
const SKIP_SIZE_THRESHOLD = 500 * 1024;

const ACCEPTED_TYPES = new Set(["image/jpeg", "image/png", "image/webp"]);

export async function compressImage(file: File): Promise<File> {
  if (ACCEPTED_TYPES.has(file.type) && file.size < SKIP_SIZE_THRESHOLD) {
    return file;
  }

  let bitmap: ImageBitmap;
  try {
    // imageOrientation: "from-image" — bắt buộc chỉ định rõ vì 1 số engine cũ
    // default "none" (bỏ qua EXIF) → ảnh chụp dọc từ điện thoại bị nén ra xoay ngang.
    bitmap = await createImageBitmap(file, { imageOrientation: "from-image" });
  } catch {
    // Trình duyệt không decode được — thường do HEIC/HEIF (Chrome không hỗ trợ).
    // Format đã được backend chấp nhận thì giữ nguyên gốc, để vẫn upload được.
    if (ACCEPTED_TYPES.has(file.type)) return file;
    throw new Error(
      `Không xử lý được ảnh "${file.name}" — có thể do định dạng HEIC/HEIF. Vui lòng đổi cài đặt camera sang JPEG hoặc chọn ảnh khác.`
    );
  }

  const scale = Math.min(1, MAX_DIMENSION / Math.max(bitmap.width, bitmap.height));
  const w = Math.round(bitmap.width * scale);
  const h = Math.round(bitmap.height * scale);

  const canvas = document.createElement("canvas");
  canvas.width = w;
  canvas.height = h;
  const ctx = canvas.getContext("2d");
  if (!ctx) {
    bitmap.close();
    return file;
  }
  ctx.drawImage(bitmap, 0, 0, w, h);
  bitmap.close();

  const blob: Blob | null = await new Promise((resolve) =>
    canvas.toBlob(resolve, "image/jpeg", JPEG_QUALITY)
  );
  if (!blob || blob.size >= file.size) return file; // nén không hiệu quả → giữ gốc

  const newName = file.name.replace(/\.\w+$/, "") + ".jpg";
  return new File([blob], newName, { type: "image/jpeg" });
}

/** Nén tuần tự (không parallel) để tránh nhiều ảnh lớn decode đồng thời ngốn RAM. */
export async function compressImages(
  files: File[],
  onProgress?: (current: number, total: number) => void
): Promise<{ compressed: File[]; errors: string[] }> {
  const compressed: File[] = [];
  const errors: string[] = [];
  for (let i = 0; i < files.length; i++) {
    onProgress?.(i, files.length);
    try {
      compressed.push(await compressImage(files[i]));
    } catch (e: any) {
      errors.push(e?.message || `Lỗi xử lý ảnh "${files[i].name}"`);
    }
  }
  onProgress?.(files.length, files.length);
  return { compressed, errors };
}
